addons/l10n_id_efaktur/models/account_move.py

```diff
@@ -122,250 +130,72 @@ def reset_efaktur(self):
         return True
 
     def download_csv(self):
-        action = {
-            'type': 'ir.actions.act_url',
-            'url': "web/content/?model=ir.attachment&id=" + str(self.l10n_id_attachment_id.id) + "&filename_field=name&field=datas&download=true&name=" + self.l10n_id_attachment_id.name,
-            'target': 'self'
-        }
-        return action
+        return self.l10n_id_efaktur_document.action_download()
 
     def download_efaktur(self):
         """Collect the data and execute function _generate_efaktur."""
         for record in self:
             if record.state == 'draft':
                 raise ValidationError(_('Could not download E-faktur in draft state'))
-
+            if not record.country_code == 'ID':
+                raise ValidationError(_("E-faktur is only available on invoices under Indonesian companies"))
+            if not record.partner_id.commercial_partner_id.l10n_id_pkp:
+                raise ValidationError(_("E-faktur is only available for taxable customers"))
+            if not record.move_type == 'out_invoice':
+                raise ValidationError(_("E-faktur is only available for invoices"))
+            if not record.line_ids.tax_ids:
+                raise ValidationError(_('E-faktur is not available for invoices without any taxes.'))
             if not record.l10n_id_tax_number:
-                if not record.l10n_id_need_kode_transaksi:
-                    raise ValidationError(_('E-faktur is not available for invoices without any taxes.'))
-                raise ValidationError(_('Connect %(move_number)s with E-faktur to download this report', move_number=record.name))
+                raise ValidationError(_("Please reset %(move_number)s to draft and post it again to generate the eTax number", move_number=record.name))
+
+        # Should prevent users from generating e-Faktur document on invoices across multi-company.
+        # Allowing it will cause issues on the invoice/eFaktur document record rule
+        if len(self.company_id) > 1:
+            raise UserError(_("You are not allowed to generate e-Faktur document from invoices coming from different company"))
+
+        # All invoices in self have no documents; we can create a new one for them.
+        # Or all invoices in self have a document, but it's the same one. Special use case but we allow downloading it.
+        if not self.l10n_id_efaktur_document:
+            self.l10n_id_efaktur_document = self.env['l10n_id_efaktur.document'].create({'invoice_ids': self.ids})
```

Comment:
Select two companies in the selector => you see all invoices of both.
Your "active" company is company A.
Select two invoices from company B, download efaktur => Won't it create a document with company A as company_id since it's the one in the env?

To be safe:
```suggestion
            self.l10n_id_efaktur_document = self.env['l10n_id_efaktur.document'].create({
                'invoice_ids': self.ids,
                'company_id': self.company_id,
            })
```